### ISSUE TITLE
`Not enough arguments` with splatted kwargs

### ISSUE THREAD

**Issue by ojab** [castwide/solargraph#396]
```ruby
def xxx(from:, to:)
end

xxx(**{from: 1, to: 2})
```
```
ojab [ /tmp ]$ solargraph typecheck 1.rb
/tmp/1.rb:4 - Not enough arguments to #xxx
1 problem found.
```

**Comment by castwide** [comment]
Confirmed. It works as expected with `xxx(from: 1, to: 2)` but misinterprets the splatted hash. I'll add it to the roadmap.

**Comment by castwide** [comment]
Released in v0.40.1.

**Comment by ojab** [comment]
Looks like the issue still appears in a slightly more complicated case:
```ruby
def xxx(from:, to:)
end

params = {from: 1, to: 2}

xxx(**params)
```

```
/tmp/sg/sg.rb:7 - Not enough arguments to #xxx
1 problem found.
```

**Comment by castwide** [comment]
Thanks. In this case, I probably won't try to validate the value of `params` in the method call; instead, the type checker will see that the kwargs are coming from a variable and ignore it. I'll at least make it stop throwing that `not enough arguments` error, though.

**Comment by castwide** [comment]
Updated in v0.40.2.

**Comment by amake** [comment]
I have now the following error if the kwargs have default arguments:

```sh
% solargraph typecheck test.rb
/Users/amake/test/test.rb:6 - Too many arguments to #xxx
1 problem found.
```

test.rb

```ruby
def xxx(from: nil, to: nil)
end

params = {from: 1, to: 2}

xxx(**params)
```

**Comment by castwide** [comment]
Fixed in v0.40.3.

**Comment by wagenet** [comment]
There still seems to be a related issue if you have both splatted args and kwargs.

```rb
  def self.day_range(range, *args, **kwargs)
    new([:day_range, range], *args, **kwargs) # Not enough arguments to MyClass.new
  end

  def initialize(timeframe, scope = nil, now: Time.now)
     # snipped
  end
```

**Comment by wagenet** [comment]
This also complains:

```rb
def by_auth_params(token:, name:, environment:, agent_reporting_env: false)
  # snipped
end

# Not enough arguments to by_auth_params
by_auth_params(
  token: @token,
  agent_reporting_env: reporting_env?,
  **component_attributes
)
```

**Comment by castwide** [comment]
The above two issues with kwrestargs are fixed in v0.43.0.
